Docker example using bun

JakobKlotz opened the pull request:
### What?

Expanded the `with-docker` example for usage with `bun` and added build instructions to the README.

### Why?

The existing example supports npm, yarn and pnpm, but not bun. 

### How?

- Add a separate Dockerfile.bun placed next to the existing Dockerfile, which is more sensible for the existing solution.
- Dockerfile.bun uses the official oven/bun base image (different runtime/tooling than node:18-alpine), so a separate file should be clearer.
- Existing Docker commands in the README remain unchanged, because Docker still uses the existing Dockerfile by default.
- The bun example can be built with: `docker build -f Dockerfile.bun -t nextjs-docker .`

Related to #78465, #82998

JakobKlotz committed:
docs (examples): add `Dockerfile` for `bun`

JakobKlotz committed:
docs (examples): add instructions for `bun`

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 6cae3838667b6fd589395d8e697ae3032c66eb4b

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

JakobKlotz committed:
docs (examples): fix multi-line ENV statement

icyJoseph commented:
Hi, could you try to address the bot comments?

JakobKlotz commented on examples/with-docker/Dockerfile.bun:
```diff
@@ -0,0 +1,61 @@
+    PORT=3000 \
+    HOSTNAME="0.0.0.0"
+
+RUN addgroup --system --gid 1001 nodejs && \
+    adduser --system --uid 1001 nextjs
```
The `addgroup` and `adduser` commands are also available on debian, "they are friendlier front ends to the low level tools like useradd, groupadd". (debian manual - source [1](https://manpages.debian.org/jessie/adduser/addgroup.8.en.html#DESCRIPTION))

JakobKlotz commented on examples/with-docker/Dockerfile.bun:
```diff
@@ -0,0 +1,60 @@
+
+# Install dependencies with bun
+FROM base AS deps
+COPY package.json bun.lock* ./
```
There's no `node_modules.bun` lockfile.
bun uses the text based `bun.lock` by default, versions prior to `1.2` used a binary `bun.lockb` ([bun docs](https://bun.com/docs/install/lockfile)). So the wildcard `bun.lock*` covers both.

JakobKlotz commented on examples/with-docker/Dockerfile.bun:
```diff
@@ -0,0 +1,60 @@
+    # Uncomment the following line in case you want to disable telemetry during runtime.
+    # ENV NEXT_TELEMETRY_DISABLED=1 \
+    PORT=3000 \
+    HOSTNAME="0.0.0.0"
```
Was fixed 🚀

JakobKlotz commented:
> Hi, could you try to address the bot comments?

Sure, addressed them and left some comments for clarification!

icyJoseph committed:
Merge branch 'canary' into docker-examples

JakobKlotz committed:
docs (examples): remove creation of next.config.js for more consistency

Existing docker example within the same directory does not create a next.config.js for standalone output (within the Dockerfile), hence it's now more consistent.

Co-authored-by: vercel[bot] <35613825+vercel[bot]@users.noreply.github.com>

icyJoseph commented on examples/with-docker/Dockerfile.bun:
```diff
@@ -0,0 +1,54 @@
+
+EXPOSE 3000
+
+CMD ["node", "server.js"]
```
Shouldn't this be `bun`?

JakobKlotz replied:
You're right, fixed that! 😄

JakobKlotz committed:
fix (examples): replace `node` with `bun` as runtime

JakobKlotz commented on examples/with-docker/Dockerfile.bun:
```diff
@@ -0,0 +1,54 @@
+
+# Install dependencies with bun
+FROM base AS deps
+COPY package.json bun.lock* ./
```
`bun` uses the text based `bun.lock` by default (see above comment).

icyJoseph commented on examples/with-docker/README.md:
```diff
@@ -43,10 +54,10 @@ This will build the project as a standalone app inside the Docker image.
+2. Run `gcloud auth login` to log in to your account.
+3. [Create a new project](https://cloud.google.com/run/docs/quickstarts/build-and-deploy) in Google Cloud Run (e.g. `nextjs-docker`). Ensure billing is turned on.
+4. Build your container image using Cloud Build: `gcloud builds submit --tag gcr.io/PROJECT-ID/helloworld --project PROJECT-ID`. This will also enable Cloud Build for your project.
+5. Deploy to Cloud Run: `gcloud run deploy --image gcr.io/PROJECT-ID/helloworld --project PROJECT-ID --platform managed --allow-unauthenticated`. Choose a region of your choice.
```
Just a nit, this was ok - 

Markdown engines see a numbered list, so they put an `ol` w/ `li` children, and the numbering is handled by the browser

1.
1.

The number is almost just a decoration, this is the output from the markdown:

```
<ol>
<li></li>
<li></li>
</ol>
```

JakobKlotz replied:
Ah sorry, I didn't know that, that's quite handy. Thanks for the clarification!

JakobKlotz committed:
docs (examples): revert list numbering change

icyJoseph commented on examples/with-docker/Dockerfile.bun:
```diff
@@ -0,0 +1,54 @@
+
+EXPOSE 3000
+
+CMD ["bun", "run", "server.js"]
```
Final thing, is it better to use the so called naked command? `bun server.js` - I guess it is a noop but `bun run` can also invoke a package.json script right?

JakobKlotz replied:
You're right, how about using the naked command including the path ([bun resolution order](https://bun.com/docs/cli/run#resolution-order))? That removes any ambiguity. 

```suggestion
CMD ["bun", "./server.js"]
```

icyJoseph replied:
Mmm sure - let's go with that and we should be ready to go

icyJoseph commented on examples/with-docker/README.md:
```diff
@@ -18,17 +18,28 @@ yarn create next-app --example with-docker nextjs-docker
 ### In existing projects
 
-To add support for Docker to an existing project, just copy the [`Dockerfile`](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile) into the root of the project and add the following to the `next.config.js` file:
+To add support for Docker to an existing project, just copy the [`Dockerfile`](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile) into the root of the project (or [`Dockerfile.bun`](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile.bun) if you're using bun) and add the following to the `next.config.js` file:
```
Wdyt?

```suggestion
To add Docker support, copy [`Dockerfile`](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile) to the project root. If using Bun, copy [`Dockerfile.bun`](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile.bun) instead. Then add the following to next.config.js:
```

JakobKlotz replied:
Sounds way better 👍🏽

JakobKlotz committed:
Update examples/with-docker/README.md

Co-authored-by: Joseph <sephxd1234@gmail.com>

JakobKlotz committed:
Update examples/with-docker/Dockerfile.bun

icyJoseph committed:
Merge branch 'canary' into docker-examples

icyJoseph commented on examples/with-docker/README.md:
```diff
@@ -4,7 +4,7 @@ This examples shows how to use Docker with Next.js based on the [deployment docu
 ## How to use
 
-Execute [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app) with [npm](https://docs.npmjs.com/cli/init), [Yarn](https://yarnpkg.com/lang/en/docs/cli/create/), or [pnpm](https://pnpm.io) to bootstrap the example:
+Execute [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app) with [npm](https://docs.npmjs.com/cli/init), [Yarn](https://yarnpkg.com/lang/en/docs/cli/create/), [pnpm](https://pnpm.io) or [bun](https://bun.com/) to bootstrap the example:
```
```suggestion
Execute [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app) with [npm](https://docs.npmjs.com/cli/init), [Yarn](https://yarnpkg.com/lang/en/docs/cli/create/), [pnpm](https://pnpm.io) or [bun](https://bun.sh/docs/cli/bun-create) to bootstrap the example:
```

icyJoseph committed:
Update examples/with-docker/README.md
